index.js

```diff
@@ -54,69 +54,126 @@ class PriorityQueue {
 		}
+		if (!(typeof opts.intervalCap === 'number' && opts.intervalCap >= 1)) {
+			throw new TypeError(`Expected \`intervalCap\` to be a number from 1 and up, got \`${opts.intervalCap}\` (${typeof opts.intervalCap})`);
+		}
+		if (!(typeof opts.intervalLength === 'number' && Number.isFinite(opts.intervalLength) && opts.intervalLength >= 0)) {
+			throw new TypeError(`Expected \`intervalLength\` to be a finite number >= 0, got \`${opts.intervalLength}\` (${typeof opts.intervalLength})`);
+		}
+
+		this._concurrentCountIsNewIntervalCount = opts.mustFinishDuringInterval;
+		this._isPaused = opts.autoStart === false;
 
 		this.queue = new opts.queueClass(); // eslint-disable-line new-cap
 		this._queueClass = opts.queueClass;
-		this._pendingCount = 0;
-		this._concurrency = opts.concurrency;
-		this._isPaused = opts.autoStart === false;
+
+		this._concurrentCount = 0;
+		this._concurrentCap = opts.concurrency;
+
+		this._isIntervalIgnored = opts.intervalCap === Infinity || opts.intervalLength === 0;
+		this._intervalCount = 0;
+		this._intervalCap = opts.intervalCap;
+		this._intervalTime = opts.intervalLength;
+		this._intervalId = null;
+
 		this._resolveEmpty = () => {};
 		this._resolveIdle = () => {};
 	}
 
-	_next() {
-		this._pendingCount--;
+	get _doesIntervalAllowAnother() {
+		return this._isIntervalIgnored || this._intervalCount < this._intervalCap;
+	}
 
-		if (this.queue.size > 0) {
-			if (!this._isPaused) {
-				this.queue.dequeue()();
-			}
-		} else {
-			this._resolveEmpty();
-			this._resolveEmpty = () => {};
+	get _areAnyQueued() {
+		return this.queue.size > 0;
+	}
+
+	get _doesConcurrentAllowAnother() {
+		return this._concurrentCount < this._concurrentCap;
+	}
+
+	_onIndividualCompletion() {
+		this._concurrentCount--;
+		this._tryToStartAnother();
+	}
+
+	_tryToStartAnother() {
+		if (!this._areAnyQueued) {
+			this._resolvePromises();
+			return false;
+		}
+		if (!this._isPaused && this._doesIntervalAllowAnother && this._doesConcurrentAllowAnother) {
+			this.queue.dequeue()();
+			this._initializeIntervalIfNeeded();
+			return true;
+		}
+		return false;
+	}
+
+	_initializeIntervalIfNeeded() {
+		if (this._isIntervalIgnored || this._intervalId !== null) {
+			return;
+		}
+		this._intervalId = setInterval(() => this._onInterval(), this._intervalTime);
+	}
+
+	_resolvePromises() {
+		this._resolveEmpty();
+		this._resolveEmpty = () => {};
+
+		if (this._concurrentCount === 0) {
+			this._resolveIdle();
+			this._resolveIdle = () => {};
+		}
+	}
 
-			if (this._pendingCount === 0) {
-				this._resolveIdle();
-				this._resolveIdle = () => {};
-			}
+	_onInterval() {
+		if (this._intervalCount === 0 && this._concurrentCount === 0) {
+			clearInterval(this._intervalId);
+			this._intervalId = null;
 		}
+		this._intervalCount = (this._concurrentCountIsNewIntervalCount) ? this._concurrentCount : 0;
```

Comment:
`_intervalCount` will never decrease except for here when the interval time ends so it needs to be set to either 0 or the `concurrentCount`. When `_onInterval()` is called it will check if the JavaScript Interval is still needed, reset the counter, and then start running promises. 
It is a bit redundant with the preceding if block so it could be placed in an else block. The `while (this._tryToStartAnother()) {}` needs to be called after either so a new set of promises can run.